File: src/scanner/lines.js
/**
 * Precomputes line-start byte offsets once per file so individual matches can
 * resolve a line number in O(log n) instead of re-slicing and re-splitting
 * the whole file per match (which is O(file size) per match — quadratic
 * overall on files with many matches).
 */
export function buildLineIndex(content) {
  const offsets = [0];
  for (let i = 0; i < content.length; i++) {
    if (content.charCodeAt(i) === 10 /* \n */) offsets.push(i + 1);
  }
  return offsets;
}

/**
 * Returns the 1-indexed line number containing byte offset `index`.
 */
export function lineNumberAt(lineOffsets, index) {
  let lo = 0;
  let hi = lineOffsets.length - 1;
  while (lo < hi) {
    const mid = (lo + hi + 1) >> 1;
    if (lineOffsets[mid] <= index) {
      lo = mid;
    } else {
      hi = mid - 1;
    }
  }
  return lo + 1;
}

File: python/src/rune/scanner/lines.py
"""
Precomputes line-start offsets once per file so individual regex matches
can resolve a line number in O(log n) via bisect instead of re-slicing and
re-splitting the whole file per match (which is O(file size) per match —
quadratic overall on files with many matches). Mirrors scanner/lines.js.
"""

import bisect


def build_line_index(content: str) -> list[int]:
    offsets = [0]
    for i, ch in enumerate(content):
        if ch == "\n":
            offsets.append(i + 1)
    return offsets


def line_number_at(line_offsets: list[int], index: int) -> int:
    """Returns the 1-indexed line number containing character offset `index`."""
    pos = bisect.bisect_right(line_offsets, index) - 1
    return pos + 1
